devito/types/basic.py

```diff
@@ -1661,3 +1662,35 @@ class IrregularFunctionInterface:
     @property
     def nbytes_max(self):
         raise NotImplementedError
+
+
+class LocalType(Basic):
+    """
+    This is the abstract base class for local types, which are
+    generated by the compiler in C rather than in Python.
+
+    Notes
+    -----
+    Subclasses should setup `_liveness`.
+    """
+
+    is_LocalType = True
+
+    @property
+    def liveness(self):
+        return self._liveness
+
+    @property
+    def _mem_internal_eager(self):
+        return self._liveness == 'eager'
+
+    @property
+    def _mem_internal_lazy(self):
+        return self._liveness == 'lazy'
+
+    """
+    A modifier added to the subclass C declaration when it appears
+    in a function signature. For example, a subclass might define `_C_modifier = '&'`
+    to impose pass-by-reference semantics.
+    """
+    _C_modifier = None
```

Comment:
FWIW, with the advent of CustomDType this might not be necessary anymore. Have a look maybe when you got time (not for this PR)